File: frontend/src/types/application.ts
export enum ApplicationStatus {
  PLANNED = 'planned',
  APPLIED = 'applied',
  INTERVIEW = 'interview',
  REJECTED = 'rejected',
  OFFER = 'offer'
}

export enum ApplicationPriority {
  HIGH = 'high',
  MEDIUM = 'medium',
  LOW = 'low'
}

export enum EmploymentType {
  FULL_TIME = 'full-time',
  PART_TIME = 'part-time',
  CONTRACT = 'contract',
  INTERNSHIP = 'internship'
}

export enum RemotePolicy {
  ON_SITE = 'on-site',
  HYBRID = 'hybrid',
  REMOTE = 'remote'
}

export interface Application {
  id: number;
  company: string;
  position_title: string;
  position_level?: string;
  department?: string;
  employment_type?: EmploymentType;
  salary_min?: number;
  salary_max?: number;
  currency: string;
  location?: string;
  remote_policy?: RemotePolicy;
  status: ApplicationStatus;
  priority: ApplicationPriority;
  source?: string;
  source_url?: string;
  notes?: string;
  tags?: string[];
  created_at: string;
  planned_date?: string;
  applied_date?: string;
  interview_date?: string;
  status_updated_at?: string;
  final_status?: string;
  rejection_reason?: string;
  feedback?: string;
  // Linked resources
  job_description_id: number;
  resume_id?: number;
  tailored_resume_id?: number;
  cover_letter_id?: number;
  outreach_draft_id?: number;
}

export interface ApplicationCreate {
  company: string;
  position_title: string;
  position_level?: string;
  department?: string;
  employment_type?: EmploymentType;
  salary_min?: number;
  salary_max?: number;
  currency?: string;
  location?: string;
  remote_policy?: RemotePolicy;
  priority?: ApplicationPriority;
  source?: string;
  source_url?: string;
  notes?: string;
  tags?: string[];
  planned_date?: string;
  // Linked resources
  job_description_id: number;
  resume_id?: number;
  tailored_resume_id?: number;
  cover_letter_id?: number;
  outreach_draft_id?: number;
}

export interface ApplicationUpdate {
  company?: string;
  position_title?: string;
  position_level?: string;
  department?: string;
  employment_type?: EmploymentType;
  salary_min?: number;
  salary_max?: number;
  currency?: string;
  location?: string;
  remote_policy?: RemotePolicy;
  status?: ApplicationStatus;
  priority?: ApplicationPriority;
  source?: string;
  source_url?: string;
  notes?: string;
  tags?: string[];
  planned_date?: string;
  applied_date?: string;
  interview_date?: string;
  final_status?: string;
  rejection_reason?: string;
  feedback?: string;
  // Linked resources
  resume_id?: number;
  tailored_resume_id?: number;
  cover_letter_id?: number;
  outreach_draft_id?: number;
}

export interface ApplicationSummary {
  id: number;
  company: string;
  position_title: string;
  status: ApplicationStatus;
  priority: ApplicationPriority;
  location?: string;
  created_at: string;
  applied_date?: string;
  interview_date?: string;
  final_status?: string;
}

export interface ApplicationDashboard {
  applications: ApplicationSummary[];
  total_count: number;
  status_counts: Record<string, number>;
  priority_counts: Record<string, number>;
  recent_activity: ApplicationSummary[];
}

export interface ApplicationStatistics {
  total_applications: number;
  recent_applications: number;
  success_rate: number;
  status_distribution: Record<string, number>;
  priority_distribution: Record<string, number>;
  offers_count: number;
}

export interface ApplicationFilters {
  status?: string;
  priority?: string;
  company?: string;
  search?: string;
  skip?: number;
  limit?: number;
}

export interface ApplicationStatusOption {
  value: ApplicationStatus;
  label: string;
  description: string;
}

export interface ApplicationPriorityOption {
  value: ApplicationPriority;
  label: string;
  description: string;
}

export interface ApplicationSourceOption {
  value: string;
  label: string;
  description: string;
}

export interface ApplicationFormData {
  company: string;
  position_title: string;
  position_level: string;
  department: string;
  employment_type: string;
  salary_min: string;
  salary_max: string;
  currency: string;
  location: string;
  remote_policy: string;
  priority: ApplicationPriority;
  source: string;
  source_url: string;
  notes: string;
  tags: string[];
  planned_date: string;
  job_description_id: number;
  resume_id?: number;
  tailored_resume_id?: number;
  cover_letter_id?: number;
  outreach_draft_id?: number;
}

File: backend/app/models/application.py
from sqlalchemy import Column, Integer, String, DateTime, Text, ForeignKey, JSON, Enum
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
import enum

from app.core.database import Base


class ApplicationStatus(str, enum.Enum):
    PLANNED = "planned"
    APPLIED = "applied"
    INTERVIEW = "interview"
    REJECTED = "rejected"
    OFFER = "offer"


class Application(Base):
    __tablename__ = "applications"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    resume_id = Column(Integer, ForeignKey("resumes.id"), nullable=True)
    tailored_resume_id = Column(Integer, ForeignKey("tailored_resumes.id"), nullable=True)
    cover_letter_id = Column(Integer, ForeignKey("cover_letters.id"), nullable=True)
    outreach_draft_id = Column(Integer, ForeignKey("outreach_drafts.id"), nullable=True)
    job_description_id = Column(Integer, ForeignKey("job_descriptions.id"), nullable=False)
    company = Column(String(255), nullable=False)
    position_title = Column(String(255), nullable=False)
    position_level = Column(String(50), nullable=True)  # entry, junior, senior, lead
    department = Column(String(100), nullable=True)
    employment_type = Column(String(50), nullable=True)  # full-time, part-time, contract
    salary_min = Column(Integer, nullable=True)
    salary_max = Column(Integer, nullable=True)
    currency = Column(String(3), default="USD")
    location = Column(String(255), nullable=True)
    remote_policy = Column(String(50), nullable=True)  # on-site, hybrid, remote
    status = Column(Enum(ApplicationStatus), default=ApplicationStatus.PLANNED, nullable=False)
    priority = Column(String(10), default="medium")  # high, medium, low
    source = Column(String(50), nullable=True)  # LinkedIn, Indeed, Company Website, Referral
    source_url = Column(String(500), nullable=True)
    notes = Column(Text, nullable=True)
    tags = Column(JSON, nullable=True)  # Array of tags
    # Timeline fields
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    planned_date = Column(DateTime(timezone=True), nullable=True)
    applied_date = Column(DateTime(timezone=True), nullable=True)
    interview_date = Column(DateTime(timezone=True), nullable=True)
    status_updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    # Outcome fields
    final_status = Column(String(50), nullable=True)  # rejected, offer, withdrawn
    rejection_reason = Column(String(255), nullable=True)  # not qualified, culture fit, position filled
    offer_details = Column(JSON, nullable=True)  # Offer details if received
    feedback = Column(Text, nullable=True)  # Company feedback if provided

    # Relationships
    user = relationship("User", back_populates="applications")
    resume = relationship("Resume", back_populates="applications")
    tailored_resume = relationship("TailoredResume", back_populates="applications")
    cover_letter = relationship("CoverLetter", back_populates="applications")
    outreach_draft = relationship("OutreachDraft", back_populates="applications")
    job_description = relationship("JobDescription", back_populates="applications")
    outreach_drafts = relationship("OutreachDraft", back_populates="application")
    interview_prep = relationship("InterviewPrep", back_populates="application")

    def __repr__(self):
        return f"<Application(id={self.id}, company={self.company}, status={self.status})>"
